crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+//! State relating to validating a WebAssembly component.
+
+use super::{check_max, core::Module};
+use crate::{
+    limits::*,
+    validator::core::{EntityType, ModuleTypeSpace},
+    BinaryReaderError, CanonicalOption, ExternalKind, FuncType, GlobalType, MemoryType, Result,
+    TableType, Type, WasmFeatures,
+};
+use std::{collections::HashMap, rc::Rc, sync::Arc};
+
+fn check_options(options: &[CanonicalOption], offset: usize) -> Result<()> {
+    fn display(option: CanonicalOption) -> &'static str {
+        match option {
+            CanonicalOption::UTF8 => "utf8",
+            CanonicalOption::UTF16 => "utf16",
+            CanonicalOption::CompactUTF16 => "compact-utf16",
+            CanonicalOption::Into(_) => "into",
+        }
+    }
+
+    let mut encoding = None;
+    let mut into = None;
+
+    for option in options {
+        match option {
+            CanonicalOption::UTF8 | CanonicalOption::UTF16 | CanonicalOption::CompactUTF16 => {
+                match encoding {
+                    Some(existing) => {
+                        return Err(BinaryReaderError::new(
+                            format!(
+                                "canonical option `{}` conflicts with option `{}`",
+                                display(existing),
+                                display(*option)
+                            ),
+                            offset,
+                        ))
+                    }
+                    None => encoding = Some(*option),
+                }
+            }
+            CanonicalOption::Into(i) => {
+                into = match into {
+                    None => Some(*i),
+                    Some(_) => {
+                        return Err(BinaryReaderError::new(
+                            "canonical option `into` is specified more than once",
+                            offset,
+                        ))
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Clone)]
+struct ModuleType {
+    imports: Rc<HashMap<(String, String), EntityType>>,
+    exports: Rc<HashMap<String, EntityType>>,
```

Comment:
Ok I'm gonna write the longer comment here now that I've read through everything. I personally feel like the previous approach of building up a list of types and having a layer of type indices into this layer is the way to go. That way "cloning" a type is just getting a new index to it. This giant list of types lives outside of any index space addressable by wasm itself and provides a future means by which to deduplicate if we want as well.

I'm wary of `Rc` and `Arc` for a few reasons:

* Lots of little allocations all over the place that we can't pre-allocate (unlike a vector)
* Overhead, especially in the `Arc` case, of maintaining the reference count
* Non-`Send`-ness of `Rc` which limits what can be done with component type information

Using a flat indexed list should solve all this, although it won't be quite as natural to work with in the internals I believe because it's a layer of indirection where you have to go back to a central data structure to get at the data.

I suspet though you removed the `SnapshotList` thing for a reason. Do you agree with my personal leaning that a list should be used instead of `Rc`/`Arc`?

Reply:
Yeah, I agree with the cons listed here and will change it back to a top-level type store with the various validator lists referencing things by index rather than the various `Rc`s.